tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawSetBuilder.scala

```diff
@@ -0,0 +1,139 @@
+      pairs: TlaEx*): TlaEx = {
+    // Even # of args and every other argument is NameEx
+    require(pairs.size % 2 == 0)
+    require(TlaOper.deinterleave(pairs)._1.forall { _.isInstanceOf[NameEx] })
+    simpleInstruction(TlaSetOper.map, e +: x1 +: set1 +: pairs: _*)
+  }
+
+  /** Function set constructor [fromSet -> toSet] */
+  protected def _funSet(fromSet: TlaEx, toSet: TlaEx): TlaEx = simpleInstruction(TlaSetOper.funSet, fromSet, toSet)
+
+  /** Record set constructor [ k1: v1, ... , kN: vN ]; must have at least 1 key-value pair */
+  protected def _recSet(k1: ValEx, v1: TlaEx, kvs: TlaEx*): TlaEx = {
+    // All keys must be ValEx(TlaStr(_))
+    require((k1 +: TlaOper.deinterleave(kvs)._1).forall {
+      case ValEx(_: TlaStr) => true
+      case _                => false
+    })
+
+    // Record constructors don't have a signature, so we must construct a type-computation manually
+    // This type computation cannot be pure, as it must read the string values of the record field names
+    val typeCmp: TypeComputation = { args =>
+      // Even-indexed values should be strings, odd-indexed values should be sets
+      val (keys, vals) = TlaOper.deinterleave(args)
+
+      // read a set's element type or throw
+      def getSetElemT(ex: TlaEx): TypeComputationReturn =
+        ex.typeTag match {
+          case Typed(SetT1(t)) => t
+          case other => throwMsg(s"Expected $ex in record set constructor to have a Set(_) type, found $other")
+        }
+
+      type exOrMap = Either[BuilderTypeException, SortedMap[String, TlaType1]]
+
+      // Iterate over the pairs (zip), first-to-throw determines the Left value, if any.
+      // If no Left appears, we get a record field-to-type map.
+      val keyTypeMapE: exOrMap =
+        keys
+          .zip(vals)
+          .foldLeft[exOrMap](Right(SortedMap.empty)) {
+            case (mapE, (ValEx(TlaStr(s)), ex)) => // key shape guaranteed by construction
+              for {
+                map <- mapE
+                setElemT <- getSetElemT(ex)
+              } yield map + (s -> setElemT)
+            case (_, (k, _)) => // Impossible, but need to suppress warning
+              Left(new BuilderTypeException(s"Key $k is not a string"))
+          }
+
+      keyTypeMapE.flatMap { keyTypeMap =>
+        // We already know all odd-index values are sets and all even-indexed are strings by
+        // require (or construction, in the string-argument method variant)
+        // so we can just map fromTypeTag over the args
+        OperT1(args.map { ex => TlaType1.fromTypeTag(ex.typeTag) }, SetT1(RecT1(keyTypeMap)))
+      }
+    }
+
+    BuilderUtil.buildInstruction(TlaSetOper.recSet, typeCmp, k1 +: v1 +: kvs: _*)
+  }
+
+  /**
+   * Alternate call method, where Scala strings are passed in place of ValEx(TlaStr(_))
+   * @see
+   *   _recSet[[_recSet(k1: ValEx, v1: TlaEx, kvs: TlaEx*)]]
+   */
+  protected def _recSet(kv1: (String, TlaEx), kvs: (String, TlaEx)*): TlaEx = {
```

Comment:
Same here. There should be no difference between the first pair and the other pairs.